File: src/main/java/com/giant/module/employee/entity/Department.java
package com.giant.module.employee.entity;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.List;

@Entity
@Table(name = "department", schema = "employee")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class Department {
    @Id
    @Column(name = "id", nullable = false)
    private Long departmentId;

    @Column(
            name = "department_name",
            nullable = false,
            unique = true
    )
    private String departmentName;

    @OneToMany(mappedBy = "department", fetch = FetchType.LAZY)
    private List<Team> teams;
}

File: src/main/kotlin/com/giant/module/employee/entity/Department.kt
package com.giant.module.employee.entity

import jakarta.persistence.*

@Entity
@Table(name = "department", schema = "employee")
class Department(

    @Column(name = "department_code", nullable = false, unique = true)
    var departmentCode: String,

    @Column(name = "department_name", nullable = false, unique = true)
    var departmentName: String,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    var departmentId: Long? = null
        protected set

    @OneToMany(mappedBy = "department", fetch = FetchType.LAZY)
    private val _teams: MutableList<Team> = mutableListOf()

    val teams: List<Team>
        get() = _teams.toList()

    companion object {
        fun create(
            departmentCode: String,
            departmentName: String
        ): Department {
            return Department(departmentCode, departmentName)
        }
    }
}
